### PR TITLE
wawenets python version

### PR DESCRIPTION
**Pull request by acatellierNTIA**
python implementation of wawenets software that has feature parity with Matlab and C++ versions. also implements the latest models from most recent wawenets article.

docstring and comment coverage should be pretty good.

have tested install procedure and code on a separate machine. have _not_ tested this code on Windows, yet.

test coverage is spotty, but additional unit tests should be relatively easy to write.

happy for this code to be looked at by more people, but didn't want to tag people out of nowhere.

### CODE REVIEW

File: python/README.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+```shell
+python wawenets_cli.py --help
+```
+
+and see its output:
+
+```shell
+Usage: wawenets_cli.py [OPTIONS]
+
+  the CLI interface Python WAWEnets. produces quality or intelligibility
```
Comment (svoranNTIA) [inline python/README.md:100]: Is the period after WAWEnets intentional?

File: python/README.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+- `-m 6`: WAWEnet trained using four subjective targets (mos, noi, col, dis) and seven objective targets (WB-PESQ, POLQA, STOI, PEMO, ViSQOL3 (c310), ESTOI, and SIIBGauss)<sup>[6](#f6)</sup>
+
+`-l L` specifies internal level normalization of `.wav` file contents to 26 dB below clipping.
+
+- `-l 0`: normalization off
+- `-l 1`: normalization on (Default)
+
+`-s S` specifies specifies the segment step (stride) and is an integer with value 1 or greater.  Default is `-s 48,000`. WAWEnet requires a full 3 seconds of signal to generate a result.  If a `.wav` file is longer than 3 seconds multiple results may be produced. `S` specifies the number of samples to move ahead in the vector or file when extracting the next segment. The default value of 48,000 gives zero overlap between segments. Using this default any input less than 6 sec. will produce one result, based on just the first 3 sec. A 6 sec. input will produce two results. If `-s 24,000` for example, segment overlap will be 50%, a 4.5 sec. input will produce 2 results and a 6 sec. input will produce 3 results.
```
Comment (svoranNTIA) [inline python/README.md:150]: replace "vector or" with "speech"

File: python/README.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
+# WAWEnets Python code
+
+Implements Wideband Audio Waveform Evaluation networks or WAWEnets.
+
+This WAWEnets implementation produces one or more speech quality or intelligibility values for each input speech signal without using reference speech signals.
+WAWEnets are convolutional networks and they have been trained using full-reference objective speech quality and speech intelligibility values.
```
Comment (svoranNTIA) [inline python/README.md:6]: replace "values." with "values as well as subjective scores."

File: python/wawenets/wawenets/model.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        Parameters
+        ----------
+        x : torch.Tensor
+            audio input data normalized to [-1, 1] with dimensions
+            [batch, 1, 48,000]
+
+        Returns
+        -------
+        torch.Tensor
+            WAWEnet predictions with dimention
```
Comment (svoranNTIA) [inline python/wawenets/wawenets/model.py:161]: replace "dimention" with "dimension"

**Review by svoranNTIA — APPROVED** [review]
Many Thanks Andrew!  Much of this is out of my domain, but I have submitted a few (maybe 5?) comments on human-readable portions. Also I noticed # TODO about 4 places.  Obviously  the code works without these enhancements but I have a question.   Is it fine to make code public with TODO's (potentially guiding others to things they can enhance) or do the TODOs need to be removed?  And yes, Jaden should work through this as well.  I will attend to that next.

File: python/wawenets/wawenets/stl_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,1117 @@
...
+            self._convert(wav_path, pcm_path)
+        else:
+            raise ValueError(
+                f"valid wav name: {valid_wav}:\n  {wav_path}\n"
+                f"valid pcm name: {valid_pcm}:\n  {pcm_path}\n"
+            )
+
+    def pcm_to_wav(self, pcm_path: Path, wav_path: Path, sample_rate: int):
+        """
+        converts the file given at `pcm_path` to a .wav file and writes teh result to
```
Comment (JPieper-NTIA) [inline python/wawenets/wawenets/stl_wrapper.py:215]: "teh" to "the"

File: python/wawenets/wawenets/stl_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,1117 @@
...
+            file path where the resampled raw (pcm) file should be written
+
+        Returns
+        -------
+        bool
+            whether the resampling operation was successful
+        """
+        with tempfile.NamedTemporaryFile() as temporary:
+            temp_path = Path(temporary.name)
+            target_ratio = 0.5
```
Comment (JPieper-NTIA) [inline python/wawenets/wawenets/stl_wrapper.py:663]: Should this be `1.5`?

**Review by JPieper-NTIA — APPROVED** [review]
Overall I think things look great! A couple comments address issues I experienced when installing/running on Windows, but those are mostly easy fixes. The code is well documented/commented and easy to read/follow the logic.

File: python/wawenets/speech/T000053_Q159_D401_48.wav [hunk 7]
```diff
@@ -0,0 +1 @@
+../../../speech/T000053_Q159_D401_48.wav
```
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wawenets/speech/T000053_Q159_D401_48.wav:1]: I was afraid of this. `git` on windows doesn't always play nice with symlinks: https://stackoverflow.com/a/59761201/650427

the files are not that big. I think it is probably more user friendly to just copy the files instead of expecting people to deal with their git install.

I think the only reason the `.wav` files are part of the python package to begin with is for tests. I think what I will actually do is make the tests generate a couple files synthetically and then delete them.

how does that sound?

File: python/wawenets/wawenets/inference.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    """Creates WAWEnets models, loads weights, and provides an interface
+    for inference."""
+
+    model_params = {
+        "icassp2020": {
+            "model_class": WAWEnetICASSP2020,
+            "num_targets": 1,
+            "channels": 96,
+        },
+        "wawenet2020": {"model_class": WAWEnet2020, "num_targets": 1, "channels": 96},
```
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wawenets/wawenets/inference.py:18]: I agree, this looks different than the others in proximity. I'm using [`black`](https://github.com/psf/black) as my python formatter, and this is how it comes out.

when you're using a formatter, "you might not like it, but at least you don't have to argue about it."

is this okay with you? should we use a different formatter?

File: python/wawenets/demo.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
+import subprocess
+import sys
+
```
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wawenets/demo.py:3]: it would be nice if they were totally exact! but in fact I had to relax the precision on one of the tests because the values were different on the two platforms I tested (intel/macOS and intel/ubuntu). I'm quite happy with differences that small to be honest.

I'll add a print statement somewhere saying that the values should be quite close but perhaps not exact.

File: python/wawenets/tests/test_data.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+    # same size
+    test_tensor = torch.zeros((1, 48))
+    result = padder({"sample_data": test_tensor})
+    assert result["sample_data"].shape[2] == 48
+
+
+class TestWavHandler:
+    def test_load_wav(self, wav_handler, test_wav):
+        # this is a stupid test now
```
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wawenets/tests/test_data.py:48]: we can lose the comment (once I fix that test!)—this was originally testing a different function I wrote and deleted a while back, now `._load_wav` just calls `torchaudio.load` and nothing else

File: python/wenets_env.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+dependencies:
+  - python=3.10
+  - tqdm
+  - pyyaml
+  - click
+  - pip
+  - poetry
+  - pytest
+  - pip:
+      - sox
```
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wenets_env.yaml:14]: added—after I commit my changes would you mind testing the env again?

File: python/wawenets/speech/T000053_Q159_D401_48.wav (hunk 7 above)
[the remark this answers is not in the document]
Comment (acatellierNTIA) [reply python/wawenets/speech/T000053_Q159_D401_48.wav:1]: I ended up removing one and keeping the other. 🤷

File: python/wawenets/wawenets/stl_wrapper.py (hunk 6 above)
Comment (acatellierNTIA) [reply python/wawenets/wawenets/stl_wrapper.py:663]: no, this is correct as-is. the way this works is that it upsamples the 24k samp/sec file to 48k samp/sec before calling `down_48k_to_16k` (which has its own target ratio). so for this intermediary step of upsampling by two, the target ratio should be 0.5.

File: python/wenets_env.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (JPieper-NTIA) [reply python/wenets_env.yaml:14]: Just did a fresh env/install and everything worked perfectly!

File: python/wawenets/wawenets/inference.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JPieper-NTIA) [reply python/wawenets/wawenets/inference.py:18]: love the quote and sounds good to me!
